File: dbc/verify_dbc.c
/*
 * Prints the real wire bytes can_protocol.c's pack_* functions produce for
 * a handful of fixed, known input values, one message per line as
 * "<id-hex> <dlc> <data-hex>". verify_dbc.py decodes each line against
 * dbc/bms.dbc (via cantools) and checks the result against the same known
 * inputs, so bms.dbc's signal layout is checked against this project's
 * actual pack/unpack code, not just hand-derived DBC bit-position math.
 *
 * Not built by the default `make sim`/`make test` -- only `make dbc-verify`
 * builds and runs this.
 */
#include <stdio.h>
#include "can_protocol.h"

static void print_frame(const can_frame_t *f) {
    printf("%03X %d ", f->id, f->dlc);
    for (int i = 0; i < f->dlc; i++) printf("%02X", f->data[i]);
    printf("\n");
}

int main(void) {
    bms_status_t status = {
        .pack_voltage_mv = 18465,
        .pack_current_ca = -200,
        .soc_percent_x2 = 196,
        .state = BMS_STATE_SHUTDOWN,
    };
    can_frame_t f_status;
    can_protocol_pack_status(&status, &f_status);
    print_frame(&f_status);

    cell_voltages_t cv = { .cell_mv = { 3650, 3700, 3695, 3712, 3680 } };
    can_frame_t f_cv;
    can_protocol_pack_cell_voltages(&cv, &f_cv);
    print_frame(&f_cv);

    fault_report_t fr = { .fault_bitmask = 0x0009, .severity = 4 };
    can_frame_t f_fault;
    can_protocol_pack_fault(&fr, &f_fault);
    print_frame(&f_fault);

    return 0;
}

File: dbc/verify_dbc.py
#!/usr/bin/env python3
"""
Reads verify_dbc.c's stdout (one "<id-hex> <dlc> <data-hex>" line per
message) and confirms dbc/bms.dbc decodes each real wire frame back to the
same known input values verify_dbc.c packed them from. Checks bms.dbc
against this project's actual pack code, not just hand-derived DBC bit
positions -- run via `make dbc-verify`, not directly.
"""
import sys
from pathlib import Path

import cantools

EXPECTED = {
    0x100: {
        "pack_voltage_mv": 18465,
        "pack_current_ca": -2.0,
        "soc_percent_x2": 98.0,
        "state": "SHUTDOWN",
    },
    0x101: {
        "cell_1_mv": 3650,
        "cell_2_mv": 3700,
        "cell_3_mv": 3695,
        "cell_4_mv": 3712,
        "cell_5_mv": 3680,
    },
    0x080: {
        "fault_bitmask": 9,
        "severity": "ASIL_D",
    },
}


def main() -> int:
    dbc_path = Path(__file__).parent / "bms.dbc"
    db = cantools.database.load_file(dbc_path)

    lines = [line for line in sys.stdin.read().splitlines() if line.strip()]
    if len(lines) != len(EXPECTED):
        print(f"expected {len(EXPECTED)} frames from verify_dbc, got {len(lines)}")
        return 1

    ok = True
    for line in lines:
        id_hex, dlc_str, data_hex = line.split()
        can_id = int(id_hex, 16)
        data = bytes.fromhex(data_hex)

        if can_id not in EXPECTED:
            print(f"0x{can_id:03X}: no expected values recorded in this script")
            ok = False
            continue

        msg = db.get_message_by_frame_id(can_id)
        if len(data) != int(dlc_str):
            print(f"0x{can_id:03X}: dlc mismatch between verify_dbc.c's stdout and itself: "
                  f"{dlc_str} vs {len(data)}")
            ok = False

        decoded = msg.decode(data)
        expected = EXPECTED[can_id]
        if decoded != expected:
            print(f"0x{can_id:03X} ({msg.name}): mismatch\n  dbc decoded: {decoded}\n  expected:    {expected}")
            ok = False
        else:
            print(f"0x{can_id:03X} ({msg.name}): OK -- {decoded}")

    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())
